kafka-client/src/main/java/dev/responsive/kafka/internal/metrics/ResponsiveRestoreListener.java

```diff
@@ -0,0 +1,227 @@
+ *  * you may not use this file except in compliance with the License.
+ *  * You may obtain a copy of the License at
+ *  *
+ *  *     http://www.apache.org/licenses/LICENSE-2.0
+ *  *
+ *  * Unless required by applicable law or agreed to in writing, software
+ *  * distributed under the License is distributed on an "AS IS" BASIS,
+ *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package dev.responsive.kafka.internal.metrics;
+
+import static dev.responsive.kafka.internal.metrics.ApplicationMetrics.NUM_INTERRUPTED_CHANGELOGS;
+import static dev.responsive.kafka.internal.metrics.ApplicationMetrics.NUM_INTERRUPTED_CHANGELOGS_DESCRIPTION;
+import static dev.responsive.kafka.internal.metrics.ApplicationMetrics.NUM_RESTORING_CHANGELOGS;
+import static dev.responsive.kafka.internal.metrics.ApplicationMetrics.NUM_RESTORING_CHANGELOGS_DESCRIPTION;
+import static dev.responsive.kafka.internal.metrics.StoreMetrics.RESTORE_LATENCY;
+import static dev.responsive.kafka.internal.metrics.StoreMetrics.RESTORE_LATENCY_AVG;
+import static dev.responsive.kafka.internal.metrics.StoreMetrics.RESTORE_LATENCY_DESCRIPTION;
+import static dev.responsive.kafka.internal.metrics.StoreMetrics.RESTORE_LATENCY_MAX;
+
+import java.io.Closeable;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.kafka.common.MetricName;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.metrics.Gauge;
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.metrics.stats.Avg;
+import org.apache.kafka.common.metrics.stats.CumulativeCount;
+import org.apache.kafka.common.metrics.stats.Max;
+import org.apache.kafka.streams.processor.StateRestoreListener;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Note: this is a global object which is shared across all StreamThreads and state stores in
+ * this Streams application.
+ */
+public class ResponsiveRestoreListener implements StateRestoreListener, Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(ResponsiveRestoreListener.class);
+
+  private final ResponsiveMetrics metrics;
+  private final MetricName numRestoringMetricName;
+  private final Sensor numInterruptedSensor;
+  private final Sensor restoreLatencySensor;
+
+  private final Map<TopicPartition, Long> restoringChangelogToStartMs = new HashMap<>();
+
+  private StateRestoreListener userRestoreListener;
+
+  public ResponsiveRestoreListener(final ResponsiveMetrics metrics) {
+    this.metrics = metrics;
+
+    numRestoringMetricName = metrics.metricName(
+        NUM_RESTORING_CHANGELOGS,
+        NUM_RESTORING_CHANGELOGS_DESCRIPTION,
+        metrics.applicationLevelMetric()
+    );
+    numInterruptedSensor = metrics.addSensor(NUM_INTERRUPTED_CHANGELOGS);
+    restoreLatencySensor = metrics.addSensor(RESTORE_LATENCY);
```

Comment:
doesn't this sensor need to be store-level? Otherwise the average/max metric tagged with a given store will reflect the average/max for all stores

Reply:
Ah, yeah it's a bit confusing but the actual metric is indeed store-level. We're just creating the sensor here but don't actually add any metrics until we have a state store begin restoration. If you look at `#onRestoreStart` you'll see where we register the actual metrics to this sensor, and those metrics are in the store-level metrics group/tagged with the specific state store name

Reply:
I think you still need store-level sensors in this scenario. For example, if you do some sequence like:

```sensor = metrics.addSensor(RESTORE_LATENCY)
...
sensor.addMetric(metricName(... args for store 1....), new Max())
sensor.addMetric(metricName(... args for store 2...), new Max())
...
// store 1 finishes restoring after 5 seconds)
sensor.record(5000)
...
// store 2 finishes restoring after 5.5 seconds)
sensor.record(5500)
```

Then, both the store 1 metric and store 2 metric will report 5500 as the max latency, where I think what you're going for is for the store-level metrics to report an aggregate over just the specific store. Basically there's no parameter in `sensor.record` that lets you scope the sample to a specific store, so you have to create store-level sensors.

Reply:
Ah, you're totally right, will do another quick followup PR

Reply:
Oh wait nevermind, this is addressed in the followup PR I already merged